### ISSUE TITLE
Build time warnings

### ISSUE THREAD

**Issue by jkeenan** [houseabsolute/List-SomeUtils-XS#2]
In the course of a lengthy ```cpan``` process to install modules needed to work on other repositories in this organization, I happened to notice these build-time warnings during installation of List-SomeUtils-XS (excerpt):
```
Running make for D/DR/DROLSKY/List-SomeUtils-XS-0.56.tar.gz
cp lib/List/SomeUtils/XS.pm blib/lib/List/SomeUtils/XS.pm
Running Mkbootstrap for XS ()
chmod 644 "XS.bs"
"/usr/local/bin/perl5.26.2" -MExtUtils::Command::MM -e 'cp_nonempty' -- XS.bs blib/arch/auto/List/SomeUtils/XS/XS.bs 644
"/usr/local/bin/perl5.26.2" "/usr/local/lib/perl5/5.26/ExtUtils/xsubpp"  -typemap '/usr/local/lib/perl5/5.26/ExtUtils/typemap'  XS.xs > XS.xsc
mv XS.xsc XS.c
cc -c    -DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_FORTIFY_SOURCE=2 -O2 -pipe -fstack-protector -fno-strict-aliasing    -DVERSION=\"0.56\"  -DXS_VERSION=\"0.56\" -DPIC -fPIC "-I/usr/local/lib/perl5/5.26/mach/CORE"   XS.c
XS.xs:160:7: warning: logical not is only applied to the left hand side of this
      bitwise operator [-Wlogical-not-parentheses]
                if (!(PadnameFLAGS(name_sv)) & SVf_OK)
                    ^                        ~
XS.xs:160:7: note: add parentheses after the '!' to evaluate the bitwise
      operator first
                if (!(PadnameFLAGS(name_sv)) & SVf_OK)
                    ^
                     (                               )
XS.xs:160:7: note: add parentheses around left hand side expression to silence
      this warning
                if (!(PadnameFLAGS(name_sv)) & SVf_OK)
                    ^
                    (                       )
XS.xs:1853:19: warning: using the result of an assignment as a condition without
      parentheses [-Wparentheses]
        while (he = hv_iternext(hv)) {
               ~~~^~~~~~~~~~~~~~~~~
XS.xs:1853:19: note: place parentheses around the assignment to silence this
      warning
        while (he = hv_iternext(hv)) {
                  ^
               (                   )
XS.xs:1853:19: note: use '==' to turn this assignment into an equality
      comparison
        while (he = hv_iternext(hv)) {
                  ^
                  ==
XS.xs:1862:19: warning: using the result of an assignment as a condition without
      parentheses [-Wparentheses]
        while (he = hv_iternext(hv)) {
               ~~~^~~~~~~~~~~~~~~~~
XS.xs:1862:19: note: place parentheses around the assignment to silence this
      warning
        while (he = hv_iternext(hv)) {
                  ^
               (                   )
XS.xs:1862:19: note: use '==' to turn this assignment into an equality
      comparison
        while (he = hv_iternext(hv)) {
                  ^
                  ==
3 warnings generated.
rm -f blib/arch/auto/List/SomeUtils/XS/XS.so
cc  -shared  -L/usr/local/lib/perl5/5.26/mach/CORE -lperl -L/usr/local/lib -fstack-protector-strong XS.o  -o blib/arch/auto/List/SomeUtils/XS/XS.so        
chmod 755 blib/arch/auto/List/SomeUtils/XS/XS.so
```
```
[vagrant@perlmongers-nycbug-generic-freebsd11 ~]$ perl -v | head -2 | tail -1
This is perl 5, version 26, subversion 2 (v5.26.2) built for amd64-freebsd-thread-multi
[vagrant@perlmongers-nycbug-generic-freebsd11 ~] uname -mrs
FreeBSD 11.1-RELEASE amd64
[vagrant@perlmongers-nycbug-generic-freebsd11 ~]$ cc --version
FreeBSD clang version 4.0.0 (tags/RELEASE_400/final 297347) (based on LLVM 4.0.0)
Target: x86_64-unknown-freebsd11.1
Thread model: posix
InstalledDir: /usr/bin
```
Can these be fixed?

Thank you very much.
Jim Keenan

**Comment by autarch** [comment]
I don't see these warnings with gcc, and I don't have an easy way to get a clang-built Perl installed (although there's probably some way to do this with perlbrew). Patches are definitely welcome here.

That said, the first warnings about the bitwise operator look like they're find a real bug. The other warnings about assignment are definitely _not_ a bug, so I'm not sure how to prevent the warnings. We want the code to do what it's doing there.

**Comment by autarch** [comment]
Fixed in v0.57 (at least for gcc warnings). Thanks!
